File: Yandex/Practicum/Algorightms_and_Data_Structures/1_sprint/L.Лишняя_буква/solution.py
import sys


def solution(x: str, y: str) -> None:
    """
    Time Complexity O(XlogX + YlogY)
    Memory Complexity O(1)
    """
    x_list: list[str] = sorted(x)
    y_list: list[str] = sorted(y)
    x_idx: int = 0
    y_idx: int = 0
    x_length: int = len(x)
    y_length: int = len(y)
    while x_idx < x_length and y_idx < y_length:
        if x_list[x_idx] != y_list[y_idx]:
            sys.stdout.write(y_list[y_idx] + '\n')
            return
        x_idx += 1
        y_idx += 1
    sys.stdout.write(y_list[y_idx] + '\n')


if __name__ == '__main__':
    x: str = sys.stdin.readline().rstrip()
    y: str = sys.stdin.readline().rstrip()
    solution(x, y)



import sys


def solution(x: str, y: str) -> None:
    """
    Time Complexity O(X + Y)
    Memory Complexity O({X} + {Y})
    """
    x_hashset: dict[str, int] = dict()
    y_hashset: dict[str, int] = dict()
    for letter in x:
        x_hashset[letter] = x_hashset.get(letter, 0) + 1
    for letter in y:
        y_hashset[letter] = y_hashset.get(letter, 0) + 1
    for letter in y_hashset:
        if letter not in x_hashset or y_hashset[letter] != x_hashset[letter]:
            sys.stdout.write(letter + '\n')


if __name__ == '__main__':
    x: str = sys.stdin.readline().rstrip()
    y: str = sys.stdin.readline().rstrip()
    solution(x, y)

File: Yandex/Practicum/Algorightms_and_Data_Structures/1_sprint/L.Лишняя_буква/solution.cpp
#include <iostream>
#include <string>
#include <algorithm>


void solution(std::string& x, std::string& y) {
    /*
    Time Complexity O(XlogX + YlogY)
    Memory Complexity O(1)
    */
    std::sort(x.begin(), x.end());
    std::sort(y.begin(), y.end());
    int xLength = x.size(), yLength = y.size();
    int xIdx = 0, yIdx = 0;
    while (xIdx < xLength && yIdx < yLength) {
        if (x[xIdx] != y[yIdx]) {
            std::cout << y[yIdx] << "\n";
            return;
        }
        ++xIdx;
        ++yIdx;
    }
    std::cout << y[yIdx] << "\n";
}


int main() {
    std::string x, y;
    std::cin >> x >> y;
    solution(x, y);
}



#include <iostream>
#include <string>
#include <unordered_map>


void solution(std::string& x, std::string& y) {
    /*
    Time Complexity O(X + Y)
    Memory Complexity O({X} + {Y})
    */
    std::unordered_map<char, int> xHashmap, yHashmap;
    for (char letter : x) {
        ++xHashmap[letter];
    }
    for (char letter : y) {
        ++yHashmap[letter];
    }
    for (auto& pair : yHashmap) {
        if (pair.second != xHashmap[pair.first]) {
            std::cout << pair.first << "\n";
        }
    }
}


int main() {
    std::string x, y;
    std::cin >> x >> y;
    solution(x, y);
}